diff -Naur expat-2.0.1/conftools/config.sub expat-2.0.1.nacl/conftools/config.sub
--- expat-2.0.1/conftools/config.sub	2006-09-22 07:42:54.000000000 -0700
+++ expat-2.0.1.nacl/conftools/config.sub	2009-10-19 15:28:36.005194000 -0700
@@ -715,6 +715,10 @@
 		basic_machine=i370-ibm
 		os=-mvs
 		;;
+	nacl)
+		basic_machine=i686-pc
+		os=-nacl
+		;;
 	ncr3000)
 		basic_machine=i486-ncr
 		os=-sysv4
@@ -1219,7 +1223,7 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -nacl*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
--- expat-2.0.1/Makefile.in	2007-05-09 11:38:50.000000000 -0700
+++ expat-2.0.1.nacl/Makefile.in	2011-06-22 12:20:25.538925743 -0700
@@ -103,6 +103,7 @@
 CC = @CC@
 CXX = @CXX@
 LIBTOOL = @LIBTOOL@
+LIBS = @LIBS@
 
 INCLUDES = -I$(srcdir)/lib -I.
 LDFLAGS = @LDFLAGS@
@@ -143,7 +144,7 @@
 xmlwf/codepage.o: xmlwf/codepage.c
 xmlwf/@FILEMAP@.o: xmlwf/@FILEMAP@.c
 xmlwf/xmlwf: $(XMLWF_OBJS) $(LIBRARY)
-	$(LINK_EXE) $(XMLWF_OBJS) $(LIBRARY)
+	$(LINK_EXE) $(XMLWF_OBJS) $(LIBRARY) $(LIBS)
 
 examples/elements.o: examples/elements.c
 examples/elements: examples/elements.o $(LIBRARY)
