diff --git a/configure b/configure
index 0ceb5bf..55e6f97 100755
--- a/configure
+++ b/configure
@@ -2974,18 +2974,22 @@ $as_echo "/* confdefs.h */" > confdefs.h
 # Predefined preprocessor variables.
 
 cat >>confdefs.h <<_ACEOF
+#undef PACKAGE_NAME
 #define PACKAGE_NAME "$PACKAGE_NAME"
 _ACEOF
 
 cat >>confdefs.h <<_ACEOF
+#undef PACKAGE_TARNAME
 #define PACKAGE_TARNAME "$PACKAGE_TARNAME"
 _ACEOF
 
 cat >>confdefs.h <<_ACEOF
+#undef PACKAGE_VERSION
 #define PACKAGE_VERSION "$PACKAGE_VERSION"
 _ACEOF
 
 cat >>confdefs.h <<_ACEOF
+#undef PACKAGE_STRING
 #define PACKAGE_STRING "$PACKAGE_STRING"
 _ACEOF
 
@@ -18805,10 +18809,10 @@ _ACEOF
 if ac_fn_c_try_compile "$LINENO"; then :
 
 
-if strings - conftest.$ac_objext | grep noonsees >/dev/null ; then
+if ${NACLSTRINGS} - conftest.$ac_objext | grep noonsees >/dev/null ; then
   ax_cv_c_float_words_bigendian=yes
 fi
-if strings - conftest.$ac_objext | grep seesnoon >/dev/null ; then
+if ${NACLSTRINGS} - conftest.$ac_objext | grep seesnoon >/dev/null ; then
   if test "$ax_cv_c_float_words_bigendian" = unknown; then
     ax_cv_c_float_words_bigendian=no
   else
