--- libav-9.6/libavutil/mem.c	2013-05-12 00:11:15.000000000 -0700
+++ libav-9.6.nacl/libavutil/mem.c	2013-06-04 02:16:01.000000000 -0700
@@ -48,11 +48,11 @@
 
 void *malloc(size_t size);
 void *memalign(size_t align, size_t size);
-int   posix_memalign(void **ptr, size_t align, size_t size);
 void *realloc(void *ptr, size_t size);
 void  free(void *ptr);
 
 #endif /* MALLOC_PREFIX */
+int   posix_memalign(void **ptr, size_t align, size_t size);
 
 /* You can redefine av_malloc and av_free in your project to use your
  * memory allocator. You do not need to suppress this file because the
--- libav-9.6/libavutil/x86/cpu.c	2013-05-12 00:11:15.000000000 -0700
+++ libav-9.6.nacl/libavutil/x86/cpu.c	2013-06-04 03:09:34.000000000 -0700
@@ -85,6 +85,12 @@
 }
 #endif
 
+#ifdef __native_client__
+#include "cpuid.h"
+#define cpuid __cpuid
+#define cpuid_test() 1
+#endif  /* __native_client__ */
+
 /* Function to test if multimedia instructions are supported...  */
 int ff_get_cpu_flags_x86(void)
 {
--- libav-9.6/doc/Makefile	2013-05-12 00:11:15.000000000 -0700
+++ libav-9.6.nacl/doc/Makefile	2013-06-04 03:16:20.000000000 -0700
@@ -29,7 +29,7 @@
 doc/%.html: TAG = HTML
 doc/%.html: doc/%.texi $(SRC_PATH)/doc/t2h.init $(GENTEXI)
 	$(Q)$(TEXIDEP)
-	$(M)texi2html -I doc -monolithic --init-file $(SRC_PATH)/doc/t2h.init --output $@ $<
+	$(M)texi2html -I doc -monolithic --init_file $(SRC_PATH)/doc/t2h.init --output $@ $<
 
 doc/%.pod: TAG = POD
 doc/%.pod: doc/%.texi $(GENTEXI)
