diff --git a/gdb/configure b/gdb/configure
index 5300b5a..8472ddd 100755
--- a/gdb/configure
+++ b/gdb/configure
@@ -5073,7 +5073,7 @@ fi
 gdb_host_obs=posix-hdep.o
 
 if test "${target}" = "${host}"; then
-  gdb_native=yes
+  gdb_native=no
 else
   gdb_native=no
 fi
diff --git a/gdb/gdb.c b/gdb/gdb.c
index d01dc46..88808fc 100644
--- a/gdb/gdb.c
+++ b/gdb/gdb.c
@@ -21,8 +21,9 @@
 #include <string.h>
 #include "interps.h"
 
-int
-main (int argc, char **argv)
+extern int gdb_nacl_main(int argc, char **argv);
+
+int gdb_nacl_main(int argc, char **argv)
 {
   struct captured_main_args args;
 
diff --git a/config/dfp.m4 b/config/dfp.m4
index 48683f0..93ac2e9 100644
--- a/config/dfp.m4
+++ b/config/dfp.m4
@@ -23,7 +23,7 @@ Valid choices are 'yes', 'bid', 'dpd', and 'no'.]) ;;
     powerpc*-*-linux* | i?86*-*-linux* | x86_64*-*-linux* | s390*-*-linux* | \
     i?86*-*-gnu* | \
     i?86*-*-mingw* | x86_64*-*-mingw* | \
-    i?86*-*-cygwin* | x86_64*-*-cygwin*)
+    i?86*-*-cygwin* | x86_64*-*-cygwin* | *nacl*)
       enable_decimal_float=yes
       ;;
     *)
diff --git a/gdb/Makefile.in b/gdb/Makefile.in
index be30dfd..50ccac8 100644
--- a/gdb/Makefile.in
+++ b/gdb/Makefile.in
@@ -115,10 +115,12 @@ MAKEINFO_CMD = $(MAKEINFO) $(MAKEINFOFLAGS) $(MAKEINFO_EXTRA_FLAGS)
 MAKEHTML = $(MAKEINFO_CMD) --html
 MAKEHTMLFLAGS =
 
+PEPPER_LIBS = -lncurses -lppapi_simple -lnacl_io -lppapi -lppapi_cpp
+
 # Set this up with gcc if you have gnu ld and the loader will print out
 # line numbers for undefined references.
 #CC_LD=gcc -static
-CC_LD=$(CC)
+CC_LD=$(CXX)
 
 # Where is our "include" directory?  Typically $(srcdir)/../include.
 # This is essentially the header file directory for the library
@@ -1221,11 +1223,11 @@ libgdb.a: $(LIBGDB_OBS)
 	$(RANLIB) libgdb.a
 
 # Removing the old gdb first works better if it is running, at least on SunOS.
-gdb$(EXEEXT): gdb.o $(LIBGDB_OBS) $(ADD_DEPS) $(CDEPS) $(TDEPLIBS)
+gdb$(EXEEXT): gdb.o gdb_pepper.o $(LIBGDB_OBS) $(ADD_DEPS) $(CDEPS) $(TDEPLIBS)
 	rm -f gdb$(EXEEXT)
-	$(CC_LD) $(INTERNAL_LDFLAGS) $(WIN32LDAPP) \
+	$(CC_LD) gdb_pepper.o $(INTERNAL_LDFLAGS) $(WIN32LDAPP) \
 		-o gdb$(EXEEXT) gdb.o $(LIBGDB_OBS) \
-		$(TDEPLIBS) $(TUI_LIBRARY) $(CLIBS) $(LOADLIBES)
+		$(TDEPLIBS) $(TUI_LIBRARY) $(CLIBS) $(LOADLIBES) $(PEPPER_LIBS)
 
 # Convenience rule to handle recursion.
 $(LIBGNU) $(GNULIB_H): all-lib
@@ -2309,7 +2311,7 @@ py-value.o: $(srcdir)/python/py-value.c
 
 # A list of all the objects we might care about in this build, for
 # dependency tracking.
-all_object_files = gdb.o $(LIBGDB_OBS) gdbtk-main.o \
+all_object_files = gdb.o gdb_pepper.o $(LIBGDB_OBS) gdbtk-main.o \
 	test-cp-name-parser.o
 
 # Ensure that generated files are created early.  Use order-only
diff --git a/libdecnumber/configure b/libdecnumber/configure
index 2720f46..39ab4c4 100755
--- a/libdecnumber/configure
+++ b/libdecnumber/configure
@@ -4616,7 +4616,7 @@ else
     powerpc*-*-linux* | i?86*-*-linux* | x86_64*-*-linux* | s390*-*-linux* | \
     i?86*-*-gnu* | \
     i?86*-*-mingw* | x86_64*-*-mingw* | \
-    i?86*-*-cygwin* | x86_64*-*-cygwin*)
+    i?86*-*-cygwin* | x86_64*-*-cygwin* | *nacl*)
       enable_decimal_float=yes
       ;;
     *)
diff --git a/libiberty/pex-unix.c b/libiberty/pex-unix.c
index addf8ee..64fc443 100644
--- a/libiberty/pex-unix.c
+++ b/libiberty/pex-unix.c
@@ -593,6 +593,9 @@ pex_unix_exec_child (struct pex_obj *obj, int flags, const char *executabl
                      int in, int out, int errdes,
                     int toclose, const char **errmsg, int *err)
 {
+#if defined(__native_client__)
+  return -1;
+#else
   pid_t pid;
 
   /* We declare these to be volatile to avoid warnings from gcc about
@@ -720,6 +723,7 @@ pex_unix_exec_child (struct pex_obj *obj, int flags, const char *executabl
 
       return pid;
     }
+#endif
 }
 #endif /* SPAWN */
 
diff --git a/gdb/charset.c b/gdb/charset.c
index 1f44e8f..03a282e 100644
--- a/gdb/charset.c
+++ b/gdb/charset.c
@@ -767,6 +767,7 @@ find_charset_names (void)
    introduction even when stdout is not a tty.  Newer versions omit
    the intro if stdout is not a tty.  */
 
+#if !defined(__native_client__)
 static int
 ignore_line_p (const char *line)
 {
@@ -790,10 +791,14 @@ ignore_line_p (const char *line)
 
   return 0;
 }
+#endif
 
 static void
 find_charset_names (void)
 {
+#if defined(__native_client__)
+  charsets = NULL;
+#else
   struct pex_obj *child;
   char *args[3];
   int err, status;
@@ -915,6 +920,7 @@ find_charset_names (void)
     }
   else
     VEC_safe_push (char_ptr, charsets, NULL);
+#endif
 }
 
 #endif /* HAVE_ICONVLIST || HAVE_LIBICONVLIST */
diff --git a/opcodes/i386-dis.c b/opcodes/i386-dis.c
index 8e55724..4b6576d 100644
--- a/opcodes/i386-dis.c
+++ b/opcodes/i386-dis.c
@@ -12635,7 +12635,8 @@ print_insn (bfd_vma pc, disassemble_info *info)
     }
 
   /* Check if the REX prefix is used.  */
-  if (rex_ignored == 0 && (rex ^ rex_used) == 0)
+  /* TODO(bradnelson): Upstream or figure out why this is a problem. */
+  if (rex && rex_ignored == 0 && (rex ^ rex_used) == 0)
     all_prefixes[last_rex_prefix] = 0;
 
   /* Check if the SEG prefix is used.  */
