diff --git a/configure.in b/configure.in
--- a/configure.in
+++ b/configure.in
@@ -920,8 +920,8 @@ AC_HELP_STRING([--enable-naclvideo], [enable the nacl video driver [[default=yes
     if test x$enable_video = xyes -a x$enable_naclvideo = xyes; then
         AC_DEFINE(SDL_VIDEO_DRIVER_NACL)
         SOURCES="$SOURCES $srcdir/src/video/nacl/*.c"
-        EXTRA_LDFLAGS="-lppapi_simple -l${NACL_CPP_LIB:-stdc++} $EXTRA_LDFLAGS"
-        SDL_LIBS="-Wl,-unacl_main -Wl,-undefined=PSUserCreateInstance -lSDLmain $SDL_LIBS -lppapi_gles2 -lcli_main -ltar -lppapi_simple -lnacl_io -lppapi_cpp -lppapi -lm -l${NACL_CPP_LIB:-stdc++}"
+        EXTRA_LDFLAGS="-lppapi_simple -l${NACL_CXX_LIB:-stdc++} $EXTRA_LDFLAGS"
+        SDL_LIBS="-Wl,-unacl_main -Wl,-undefined=PSUserMainGet -lSDLmain $SDL_LIBS -lppapi_gles2 -lcli_main -lnacl_spawn -ltar -lppapi_simple -lnacl_io -lppapi -lm -l${NACL_CXX_LIB:-stdc++}"
         SDL_CFLAGS="$SDL_CFLAGS -Dmain=SDL_main"
     fi
 
