diff -aur libxml2-2.7.8/xmlIO.c libxml2-2.7.8-nacl/xmlIO.c
--- libxml2-2.7.8/xmlIO.c	2010-10-11 23:25:32.000000000 -0700
+++ libxml2-2.7.8-nacl/xmlIO.c	2013-01-23 17:18:59.255645774 -0800
@@ -3649,10 +3649,12 @@
 	else *cur = 0;
 	ret = xmlMemStrdup(dir);
     } else {
+#ifndef __native_client__
         if (getcwd(dir, 1024) != NULL) {
 	    dir[1023] = 0;
 	    ret = xmlMemStrdup(dir);
 	}
+#endif
     }
     return(ret);
 #undef IS_XMLPGD_SEP
