diff --git a/pango/opentype/hb-common.h b/pango/opentype/hb-common.h
--- a/pango/opentype/hb-common.h
+++ b/pango/opentype/hb-common.h
@@ -29,7 +29,7 @@
 
 # ifdef HAVE_STDINT_H
 #  include <stdint.h>
-# else ifdef HAVE_INTTYPES_H
+# elif defined(HAVE_INTTYPES_H)
 # include <inttypes.h>
 # endif
 
diff --git a/pango/opentype/main.cc b/pango/opentype/main.cc
--- a/pango/opentype/main.cc
+++ b/pango/opentype/main.cc
@@ -33,6 +33,11 @@
 #include <stdlib.h>
 #include <stdio.h>
 
+#ifdef __native_client__
+extern "C" int nacl_main(int argc, char **argv);
+#define main nacl_main
+#endif
+
 int
 main (int argc, char **argv)
 {
