diff -Naur protobuf-2.5.0/configure protobuf-2.5.0.nacl/configure
--- protobuf-2.5.0/configure	2013-02-26 10:04:51.000000000 -0800
+++ protobuf-2.5.0.nacl/configure	2013-06-13 19:38:19.893895933 -0700
@@ -781,7 +781,6 @@
 CXXFLAGS
 CCC
 CXXCPP'
-ac_subdirs_all='gtest'
 
 # Initialize some variables set by options.
 ac_init_help=
@@ -16532,6 +16531,7 @@
 
 
 # Finally, execute ACTION-IF-FOUND/ACTION-IF-NOT-FOUND:
+acx_pthread_ok=yes
 if test x"$acx_pthread_ok" = xyes; then
 
 $as_echo "#define HAVE_PTHREAD 1" >>confdefs.h
@@ -16681,7 +16681,6 @@
 export CXXFLAGS
 
 
-subdirs="$subdirs gtest"
 
 
 ac_config_files="$ac_config_files Makefile src/Makefile protobuf.pc protobuf-lite.pc"
diff -Naur protobuf-2.5.0/src/Makefile.am protobuf-2.5.0.nacl/src/Makefile.am
--- protobuf-2.5.0/src/Makefile.am	2013-02-26 09:56:43.000000000 -0800
+++ protobuf-2.5.0.nacl/src/Makefile.am	2013-06-13 19:33:16.973995918 -0700
@@ -87,7 +87,7 @@
   google/protobuf/compiler/java/java_generator.h                \
   google/protobuf/compiler/python/python_generator.h
 
-lib_LTLIBRARIES = libprotobuf-lite.la libprotobuf.la libprotoc.la
+lib_LTLIBRARIES = libprotobuf-lite.la libprotobuf.la
 
 libprotobuf_lite_la_LIBADD = $(PTHREAD_LIBS)
 libprotobuf_lite_la_LDFLAGS = -version-info 8:0:0 -export-dynamic -no-undefined
diff -Naur protobuf-2.5.0/src/Makefile.in protobuf-2.5.0.nacl/src/Makefile.in
--- protobuf-2.5.0/src/Makefile.in	2013-02-26 10:04:53.000000000 -0800
+++ protobuf-2.5.0.nacl/src/Makefile.in	2013-06-13 19:33:16.973995918 -0700
@@ -563,7 +563,7 @@
   google/protobuf/compiler/java/java_generator.h                \
   google/protobuf/compiler/python/python_generator.h
 
-lib_LTLIBRARIES = libprotobuf-lite.la libprotobuf.la libprotoc.la
+lib_LTLIBRARIES = libprotobuf-lite.la libprotobuf.la
 libprotobuf_lite_la_LIBADD = $(PTHREAD_LIBS)
 libprotobuf_lite_la_LDFLAGS = -version-info 8:0:0 -export-dynamic -no-undefined
 libprotobuf_lite_la_SOURCES = \
@@ -847,7 +847,7 @@
 @HAVE_ZLIB_TRUE@zcgzip_SOURCES = google/protobuf/testing/zcgzip.cc
 @HAVE_ZLIB_TRUE@zcgunzip_LDADD = $(PTHREAD_LIBS) libprotobuf.la
 @HAVE_ZLIB_TRUE@zcgunzip_SOURCES = google/protobuf/testing/zcgunzip.cc
-all: $(BUILT_SOURCES)
+all:
 	$(MAKE) $(AM_MAKEFLAGS) all-am
 
 .SUFFIXES:
@@ -2962,14 +2962,14 @@
 	$(MAKE) $(AM_MAKEFLAGS) check-TESTS
 check: $(BUILT_SOURCES)
 	$(MAKE) $(AM_MAKEFLAGS) check-am
-all-am: Makefile $(LTLIBRARIES) $(PROGRAMS) $(DATA) $(HEADERS)
+all-am: Makefile $(LTLIBRARIES) $(DATA) $(HEADERS)
 install-binPROGRAMS: install-libLTLIBRARIES
 
 installdirs:
 	for dir in "$(DESTDIR)$(libdir)" "$(DESTDIR)$(bindir)" "$(DESTDIR)$(protodir)" "$(DESTDIR)$(includedir)"; do \
 	  test -z "$$dir" || $(MKDIR_P) "$$dir"; \
 	done
-install: $(BUILT_SOURCES)
+install:
 	$(MAKE) $(AM_MAKEFLAGS) install-am
 install-exec: install-exec-am
 install-data: install-data-am
@@ -3033,7 +3033,7 @@
 
 install-dvi-am:
 
-install-exec-am: install-binPROGRAMS install-libLTLIBRARIES
+install-exec-am: install-libLTLIBRARIES
 
 install-html: install-html-am
 
diff -Naur protobuf-2.5.0/src/google/protobuf/io/coded_stream.h protobuf-2.5.0.nacl/src/google/protobuf/io/coded_stream.h
--- protobuf-2.5.0/src/google/protobuf/io/coded_stream.h	2013-02-26 09:56:42.000000000 -0800
+++ protobuf-2.5.0.nacl/src/google/protobuf/io/coded_stream.h	2013-06-13 19:33:16.983995915 -0700
@@ -109,6 +109,7 @@
 #ifndef GOOGLE_PROTOBUF_IO_CODED_STREAM_H__
 #define GOOGLE_PROTOBUF_IO_CODED_STREAM_H__
 
+#include <climits>
 #include <string>
 #ifdef _MSC_VER
   #if defined(_M_IX86) && \

--- protobuf-2.5.0/src/google/protobuf/stubs/atomicops.h
+++ protobuf-2.5.0.nacl/src/google/protobuf/stubs/atomicops.h
@@ -176,13 +176,13 @@ GOOGLE_PROTOBUF_ATOMICOPS_ERROR
 #elif defined(__GNUC__)
 #if defined(GOOGLE_PROTOBUF_ARCH_IA32) || defined(GOOGLE_PROTOBUF_ARCH_X64)
 #include <google/protobuf/stubs/atomicops_internals_x86_gcc.h>
-#elif defined(GOOGLE_PROTOBUF_ARCH_ARM)
+#elif defined(GOOGLE_PROTOBUF_ARCH_ARM) && !defined(__native_client__)
 #include <google/protobuf/stubs/atomicops_internals_arm_gcc.h>
 #elif defined(GOOGLE_PROTOBUF_ARCH_ARM_QNX)
 #include <google/protobuf/stubs/atomicops_internals_arm_qnx.h>
 #elif defined(GOOGLE_PROTOBUF_ARCH_MIPS)
 #include <google/protobuf/stubs/atomicops_internals_mips_gcc.h>
-#elif defined(__pnacl__)
+#elif defined(__native_client__)
 #include <google/protobuf/stubs/atomicops_internals_pnacl.h>
 #else
 GOOGLE_PROTOBUF_ATOMICOPS_ERROR
