diff --git a/src/google/protobuf/compiler/command_line_interface_unittest.cc b/src/google/protobuf/compiler/command_line_interface_unittest.cc
--- a/src/google/protobuf/compiler/command_line_interface_unittest.cc
+++ b/src/google/protobuf/compiler/command_line_interface_unittest.cc
@@ -64,6 +64,10 @@
 #include <gtest/gtest.h>
 
 
+// Disable this whole test when running on Native Client since it doesn't
+// support fork/spawning new processing.
+#if !defined(__native_client__)
+
 // Disable the whole test when we use tcmalloc for "draconian" heap checks, in
 // which case tcmalloc will print warnings that fail the plugin tests.
 #if !GOOGLE_PROTOBUF_HEAP_CHECK_DRACONIAN
@@ -1772,3 +1776,4 @@ TEST_F(EncodeDecodeTest, ProtoParseError) {
 
 #endif  // !GOOGLE_PROTOBUF_HEAP_CHECK_DRACONIAN
 }  // namespace google
+#endif
diff --git a/src/google/protobuf/compiler/objectivec/objectivec_helpers.cc b/src/google/protobuf/compiler/objectivec/objectivec_helpers.cc
--- a/src/google/protobuf/compiler/objectivec/objectivec_helpers.cc
+++ b/src/google/protobuf/compiler/objectivec/objectivec_helpers.cc
@@ -28,6 +28,7 @@
 // (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+#include <climits>
 #include <fstream>
 #include <iostream>
 #include <sstream>
diff --git a/src/google/protobuf/compiler/ruby/ruby_generator_unittest.cc b/src/google/protobuf/compiler/ruby/ruby_generator_unittest.cc
--- a/src/google/protobuf/compiler/ruby/ruby_generator_unittest.cc
+++ b/src/google/protobuf/compiler/ruby/ruby_generator_unittest.cc
@@ -45,6 +45,10 @@ namespace compiler {
 namespace ruby {
 namespace {
 
+// Don't try to run ruby tests.  The source distribution doesn't seem to
+// contain the ruby test data at all
+#if !defined(__native_client__)
+
 string FindRubyTestDir(const string& file) {
   // Inspired by TestSourceDir() in src/google/protobuf/testing/googletest.cc.
   string prefix = ".";
@@ -112,6 +116,7 @@ TEST(RubyGeneratorTest, GeneratorTest) {
   EXPECT_EQ(expected_output, output);
 }
 
+#endif  // __native_client__
 }  // namespace
 }  // namespace ruby
 }  // namespace compiler
diff --git a/src/google/protobuf/io/zero_copy_stream_unittest.cc b/src/google/protobuf/io/zero_copy_stream_unittest.cc
--- a/src/google/protobuf/io/zero_copy_stream_unittest.cc
+++ b/src/google/protobuf/io/zero_copy_stream_unittest.cc
@@ -843,6 +843,9 @@ TEST_F(IoTest, FileWriteError) {
   EXPECT_EQ(EBADF, input.GetErrno());
 }
 
+// pipe() doesn't under sel_ldr.
+// TODO(sbc): Its possible we could make this work with nacl_io.
+#ifndef __native_client__
 // Pipes are not seekable, so File{Input,Output}Stream ends up doing some
 // different things to handle them.  We'll test by writing to a pipe and
 // reading back from it.
@@ -871,6 +874,7 @@ TEST_F(IoTest, PipeIo) {
     }
   }
 }
+#endif
 
 // Test using C++ iostreams.
 TEST_F(IoTest, IostreamIo) {
diff --git a/src/google/protobuf/message_unittest.cc b/src/google/protobuf/message_unittest.cc
--- a/src/google/protobuf/message_unittest.cc
+++ b/src/google/protobuf/message_unittest.cc
@@ -109,6 +109,7 @@ TEST(MessageTest, ParseFromFileDescriptor) {
   string filename = TestSourceDir() +
                     "/google/protobuf/testdata/golden_message";
   int file = open(filename.c_str(), O_RDONLY | O_BINARY);
+  ASSERT_NE(file, -1);
 
   unittest::TestAllTypes message;
   EXPECT_TRUE(message.ParseFromFileDescriptor(file));
@@ -122,6 +123,7 @@ TEST(MessageTest, ParsePackedFromFileDescriptor) {
       TestSourceDir() +
       "/google/protobuf/testdata/golden_packed_fields_message";
   int file = open(filename.c_str(), O_RDONLY | O_BINARY);
+  ASSERT_NE(file, -1);
 
   unittest::TestPackedTypes message;
   EXPECT_TRUE(message.ParseFromFileDescriptor(file));
diff --git a/src/google/protobuf/stubs/atomicops.h b/src/google/protobuf/stubs/atomicops.h
--- a/src/google/protobuf/stubs/atomicops.h
+++ b/src/google/protobuf/stubs/atomicops.h
@@ -209,7 +209,7 @@ GOOGLE_PROTOBUF_ATOMICOPS_ERROR
 #include <google/protobuf/stubs/atomicops_internals_arm_qnx.h>
 #elif defined(GOOGLE_PROTOBUF_ARCH_MIPS) || defined(GOOGLE_PROTOBUF_ARCH_MIPS64)
 #include <google/protobuf/stubs/atomicops_internals_mips_gcc.h>
-#elif defined(__native_client__)
+#elif defined(__pnacl__)
 #include <google/protobuf/stubs/atomicops_internals_pnacl.h>
 #elif (((__GNUC__ == 4) && (__GNUC_MINOR__ >= 7)) || (__GNUC__ > 4))
 #include <google/protobuf/stubs/atomicops_internals_generic_gcc.h>
diff --git a/src/google/protobuf/stubs/stringprintf_unittest.cc b/src/google/protobuf/stubs/stringprintf_unittest.cc
--- a/src/google/protobuf/stubs/stringprintf_unittest.cc
+++ b/src/google/protobuf/stubs/stringprintf_unittest.cc
@@ -54,8 +54,8 @@ TEST(StringPrintfTest, Empty) {
 }
 
 TEST(StringPrintfTest, Misc) {
-// MSVC and mingw does not support $ format specifier.
-#if !defined(_MSC_VER) && !defined(__MINGW32__)
+// MSVC, mingw and newlib don't not support $ format specifier.
+#if !defined(_MSC_VER) && !defined(__MINGW32__) && !defined(_NEWLIB_VERSION)
   EXPECT_EQ("123hello w", StringPrintf("%3$d%2$s %1$c", 'w', "hello", 123));
 #endif  // !_MSC_VER
 }
