Index: build/download_toolchains.py
===================================================================
--- build/download_toolchains.py	(revision 3514)
+++ build/download_toolchains.py	(working copy)
@@ -57,7 +57,7 @@
   parser = optparse.OptionParser()
   parser.add_option(
       '-b', '--base-url', dest='base_url',
-      default='http://build.chromium.org/buildbot/nacl_archive/nacl/toolchain',
+      default='http://commondatastorage.googleapis.com/nativeclient-archive2/toolchain',
       help='base url to download from')
   parser.add_option(
       '--x86-version', dest='x86_version',
