Index: resources/sunspider-standalone-driver.js
===================================================================
--- resources/sunspider-standalone-driver.js	(revision 27790)
+++ resources/sunspider-standalone-driver.js	(working copy)
@@ -32,7 +32,9 @@
 for (var j = 0; j < tests.length; j++) {
     var testName = "tests/" + tests[j] + ".js";
     var startTime = new Date;
-    load(testName);
+    for (var k = 0; k < 100; k++) {
+        load(testName);
+    }
     times[j] = new Date() - startTime;
     gc();
 }
