diff --git a/pyppapi/conn_stubs.c b/pyppapi/conn_stubs.c
index 36ab3f9..b59b775 100644
--- a/pyppapi/conn_stubs.c
+++ b/pyppapi/conn_stubs.c
@@ -2,6 +2,7 @@
 #include "ppapi/c/ppb_messaging.h" 
 #include "ppapi/c/pp_instance.h"
 
+#include <stddef.h>
 
 const PPB_Messaging *setup_ppapi_connection(PP_Instance *instance) __attribute__((weak));
 const PPB_Messaging *setup_ppapi_connection(PP_Instance *instance)
