diff -Naur libtomcrypt-1.17/src/headers/tomcrypt_cfg.h libtomcrypt-1.17.nacl/src/headers/tomcrypt_cfg.h
--- libtomcrypt-1.17/src/headers/tomcrypt_cfg.h	2007-05-12 07:46:25.000000000 -0700
+++ libtomcrypt-1.17.nacl/src/headers/tomcrypt_cfg.h	2010-07-23 14:58:42.000000000 -0700
@@ -70,7 +70,7 @@
 #endif
 
 /* detect amd64 */
-#if !defined(__STRICT_ANSI__) && defined(__x86_64__)
+#if !defined(__STRICT_ANSI__) && defined(__x86_64__) && !defined(__native_client__)
    #define ENDIAN_LITTLE
    #define ENDIAN_64BITWORD
    #define LTC_FAST
diff -Naur libtomcrypt-1.17/src/headers/tomcrypt_macros.h libtomcrypt-1.17.nacl/src/headers/tomcrypt_macros.h
--- libtomcrypt-1.17/src/headers/tomcrypt_macros.h	2007-05-12 07:46:25.000000000 -0700
+++ libtomcrypt-1.17.nacl/src/headers/tomcrypt_macros.h	2010-07-23 16:01:34.000000000 -0700
@@ -7,10 +7,10 @@
    typedef unsigned long long ulong64;
 #endif
 
-/* this is the "32-bit at least" data type 
- * Re-define it to suit your platform but it must be at least 32-bits 
+/* this is the "32-bit at least" data type
+ * Re-define it to suit your platform but it must be at least 32-bits
  */
-#if defined(__x86_64__) || (defined(__sparc__) && defined(__arch64__))
+#if !defined(__native_client__) && defined(__x86_64__) || (defined(__sparc__) && defined(__arch64__))
    typedef unsigned ulong32;
 #else
    typedef unsigned long ulong32;
@@ -67,7 +67,7 @@
 
 #ifdef ENDIAN_LITTLE
 
-#if !defined(LTC_NO_BSWAP) && (defined(INTEL_CC) || (defined(__GNUC__) && (defined(__DJGPP__) || defined(__CYGWIN__) || defined(__MINGW32__) || defined(__i386__) || defined(__x86_64__))))
+#if !defined(LTC_NO_BSWAP) && (defined(INTEL_CC) || (defined(__GNUC__) && (defined(__DJGPP__) || defined(__CYGWIN__) || defined(__MINGW32__) || defined(__i386__) || defined(__x86_64__) && !defined(__native_client__))))
 
 #define STORE32H(x, y)           \
 asm __volatile__ (               \
@@ -98,7 +98,7 @@
 
 
 /* x86_64 processor */
-#if !defined(LTC_NO_BSWAP) && (defined(__GNUC__) && defined(__x86_64__))
+#if !defined(LTC_NO_BSWAP) && (defined(__GNUC__) && defined(__x86_64__) && !defined(__native_client__))
 
 #define STORE64H(x, y)           \
 asm __volatile__ (               \
@@ -341,7 +341,7 @@
 
 
 /* 64-bit Rotates */
-#if !defined(__STRICT_ANSI__) && defined(__GNUC__) && defined(__x86_64__) && !defined(LTC_NO_ASM)
+#if !defined(__STRICT_ANSI__) && defined(__GNUC__) && defined(__x86_64__) && !defined(LTC_NO_ASM) && !defined(__native_client__)
 
 static inline unsigned long ROL64(unsigned long word, int i)
 {
