diff --git a/Makefile b/Makefile
--- a/Makefile
+++ b/Makefile
@@ -20,7 +20,7 @@ include Makedefs
 # Directories to make...
 #
 
-DIRS	=	cups test $(BUILDDIRS)
+DIRS	=	cups $(BUILDDIRS)
 
 
 #
diff --git a/configure b/configure
--- a/configure
+++ b/configure
@@ -5846,6 +5846,10 @@ case "$COMPONENTS" in
 		BUILDDIRS="data locale"
 		;;
 
+  core-and-image)
+    BUILDDIRS="data filter locale"
+    ;;
+
 	*)
 		as_fn_error $? "Bad build component \"$COMPONENT\" specified!" "$LINENO" 5
 		;;
diff --git a/cups/Makefile b/cups/Makefile
--- a/cups/Makefile
+++ b/cups/Makefile
@@ -79,7 +79,8 @@ LIBOBJS	=	\
 		tls.o \
 		transcode.o \
 		usersys.o \
-		util.o
+		util.o \
+		uname.o
 TESTOBJS	= \
 		testadmin.o \
 		testarray.o \
diff --git a/cups/backchannel.c b/cups/backchannel.c
--- a/cups/backchannel.c
+++ b/cups/backchannel.c
@@ -28,6 +28,7 @@
 #  include <sys/time.h>
 #endif /* WIN32 */
 
+#include <sys/select.h>
 
 /*
  * Local functions...
diff --git a/cups/usersys.c b/cups/usersys.c
--- a/cups/usersys.c
+++ b/cups/usersys.c
@@ -1044,7 +1044,9 @@ cups_read_client_conf(
 	  options = 0;
       }
 
+#ifdef HAVE_SSL
       _httpTLSSetOptions(options);
+#endif /* HAVE_SSL */
     }
   }
 
