diff -ru lua-5.2.2-tests/files.lua lua-5.2.2-tests.good/files.lua
--- lua-5.2.2-tests/files.lua	2013-03-07 10:27:28.000000000 -0800
+++ lua-5.2.2-tests.good/files.lua	2013-11-08 12:13:51.601002172 -0800
@@ -1,4 +1,5 @@
 debug = require "debug"
+nacl = true
 
 assert(type(os.getenv"PATH") == "string")
 
@@ -501,7 +502,7 @@
   x = nil; y = nil
 end
 
-if not _noposix then
+if not _noposix and not nacl then
   print("testing popen/pclose and execute")
   local tests = {
     -- command,   what,  code
@@ -511,7 +512,7 @@
     {"exit 129", "exit", 129},
     {"kill -s HUP $$", "signal", 1},
     {"kill -s KILL $$", "signal", 9},
-    {"sh -c 'kill -s HUP $$'", "exit"},
+    -- {"sh -c 'kill -s HUP $$'", "exit"},
     {'lua -e "os.exit(20, true)"', "exit", 20},
   }
   print("\n(some error messages are expected now)")
diff -ru lua-5.2.2-tests/literals.lua lua-5.2.2-tests.good/literals.lua
--- lua-5.2.2-tests/literals.lua	2013-03-07 10:27:28.000000000 -0800
+++ lua-5.2.2-tests.good/literals.lua	2013-11-07 16:29:02.226343882 -0800
@@ -1,6 +1,7 @@
 print('testing scanner')
 
 debug = require "debug"
+nacl = true
 
 
 local function dostring (x) return assert(load(x))() end
@@ -227,7 +228,7 @@
 
 
 -- testing decimal point locale
-if os.setlocale("pt_BR") or os.setlocale("ptb") then
+if not nacl and (os.setlocale("pt_BR") or os.setlocale("ptb")) then
   assert(not load("á = 3"))  -- parser still works with C locale
   assert(not load("a = (3,4)"))
   assert(tonumber("3,4") == 3.4 and tonumber"3.4" == nil)
