diff --git yajl-2.0.4/XCompile-nacl.txt yajl-2.0.4-nacl/XCompile-nacl.txt
new file mode 100644
index 0000000..dbf2917
--- /dev/null
+++ yajl-2.0.4-nacl/XCompile-nacl.txt
@@ -0,0 +1,27 @@
+# Cross-compiling requires CMake 2.6 or newer. To cross-compile, first modify
+# this file to set the proper settings and paths. Then use it from build/ like:
+# cmake .. -DCMAKE_TOOLCHAIN_FILE=../XCompile.txt \
+#          -DCMAKE_INSTALL_PREFIX=/usr/mingw32/mingw
+# If you already have a toolchain file setup, you may use that instead of this
+# file.
+
+# the name of the target operating system
+SET(CMAKE_SYSTEM_NAME Linux)
+
+# which compilers to use for C and C++
+SET(CMAKE_C_COMPILER ${NACLCC})
+SET(CMAKE_CXX_COMPILER ${NACLCXX})
+set(CMAKE_AR ${NACLAR} CACHE FILEPATH "Archiver")
+SET(_CMAKE_TOOLCHAIN_PREFIX ${NACL_CROSS_PREFIX})
+
+# here is the target environment located
+SET(CMAKE_FIND_ROOT_PATH ${NACL_TOOLCHAIN_ROOT})
+
+# adjust the default behaviour of the FIND_XXX() commands:
+# search headers and libraries in the target environment, search
+# programs in the host environment
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+set(CMAKE_REQUIRED_INCLUDES "${NACL_SDK_ROOT}/include")
+include_directories(${NACL_SDK_ROOT}/include)
diff --git yajl-2.0.4/src/CMakeLists.txt yajl-2.0.4-nacl/src/CMakeLists.txt
index e36b736..4193048 100644
--- yajl-2.0.4/src/CMakeLists.txt
+++ yajl-2.0.4-nacl/src/CMakeLists.txt
@@ -37,20 +37,22 @@ SET(LIBRARY_OUTPUT_PATH ${libDir})
 
 ADD_LIBRARY(yajl_s STATIC ${SRCS} ${HDRS} ${PUB_HDRS})
 
-ADD_LIBRARY(yajl SHARED ${SRCS} ${HDRS} ${PUB_HDRS})
+IF(BUILD_SHARED)
+  ADD_LIBRARY(yajl SHARED ${SRCS} ${HDRS} ${PUB_HDRS})
 
-#### setup shared library version number
-SET_TARGET_PROPERTIES(yajl PROPERTIES
-                      DEFINE_SYMBOL YAJL_SHARED
-                      SOVERSION ${YAJL_MAJOR}
-                      VERSION ${YAJL_MAJOR}.${YAJL_MINOR}.${YAJL_MICRO})
-
-#### ensure a .dylib has correct absolute installation paths upon installation
-IF(APPLE)
-  MESSAGE("INSTALL_NAME_DIR: ${CMAKE_INSTALL_PREFIX}/lib")
+  #### setup shared library version number
   SET_TARGET_PROPERTIES(yajl PROPERTIES
-                        INSTALL_NAME_DIR "${CMAKE_INSTALL_PREFIX}/lib")
-ENDIF(APPLE)
+                        DEFINE_SYMBOL YAJL_SHARED
+                        SOVERSION ${YAJL_MAJOR}
+                        VERSION ${YAJL_MAJOR}.${YAJL_MINOR}.${YAJL_MICRO})
+
+  #### ensure a .dylib has correct absolute installation paths upon installation
+  IF(APPLE)
+    MESSAGE("INSTALL_NAME_DIR: ${CMAKE_INSTALL_PREFIX}/lib")
+    SET_TARGET_PROPERTIES(yajl PROPERTIES
+                          INSTALL_NAME_DIR "${CMAKE_INSTALL_PREFIX}/lib")
+  ENDIF(APPLE)
+ENDIF(BUILD_SHARED)
 
 #### build up an sdk as a post build step
 
@@ -78,7 +80,9 @@ INCLUDE_DIRECTORIES(${incDir}/..)
 IF(NOT WIN32)
   # at build time you may specify the cmake variable LIB_SUFFIX to handle
   # 64-bit systems which use 'lib64'
-  INSTALL(TARGETS yajl LIBRARY DESTINATION lib${LIB_SUFFIX})
+  IF(BUILD_SHARED)
+    INSTALL(TARGETS yajl LIBRARY DESTINATION lib${LIB_SUFFIX})
+  ENDIF(BUILD_SHARED)
   INSTALL(TARGETS yajl_s ARCHIVE DESTINATION lib${LIB_SUFFIX})
   INSTALL(FILES ${PUB_HDRS} DESTINATION include/yajl)
   INSTALL(FILES ${incDir}/yajl_version.h DESTINATION include/yajl)
diff --git yajl-2.0.4/test/run_tests.sh yajl-2.0.4-nacl/test/run_tests.sh
index 55a665b..225df99 100755
--- yajl-2.0.4/test/run_tests.sh
+++ yajl-2.0.4-nacl/test/run_tests.sh
@@ -68,6 +68,8 @@ for file in cases/*.json ; do
   # parse with a read buffer size ranging from 1-31 to stress stream parsing
   while [ $iter -lt 32  ] && [ $success = "SUCCESS" ] ; do
     $testBin $allowPartials $allowComments $allowGarbage $allowMultiple -b $iter < $file > ${file}.test  2>&1
+    # sel_ldr poops out extra text to stderr.
+    sed -i -e"/^DEBUG MODE ENABLED/d" ${file}.test
     diff ${DIFF_FLAGS} ${file}.gold ${file}.test > ${file}.out
     if [ $? -eq 0 ] ; then
       if [ $iter -eq 31 ] ; then : $(( testsSucceeded += 1)) ; fi
