diff -urN libhangul-0.1.0/config.h.in libhangul-0.1.0-nacl/config.h.in
--- libhangul-0.1.0/config.h.in	2011-10-16 21:44:31.000000000 +0800
+++ libhangul-0.1.0-nacl/config.h.in	2013-02-21 12:59:43.346588775 +0800
@@ -149,6 +149,3 @@
 #ifndef __cplusplus
 #undef inline
 #endif
-
-/* Define to rpl_realloc if the replacement function should be used. */
-#undef realloc
diff -urN libhangul-0.1.0/config.sub libhangul-0.1.0-nacl/config.sub
--- libhangul-0.1.0/config.sub	2006-10-12 20:52:19.000000000 +0800
+++ libhangul-0.1.0-nacl/config.sub	2013-02-21 10:47:02.279157346 +0800
@@ -306,6 +306,14 @@
 	i*86 | x86_64)
 	  basic_machine=$basic_machine-pc
 	  ;;
+	nacl64*)
+	  basic_machine=x86_64-pc
+	  os=-nacl
+	  ;;
+	nacl*)
+	  basic_machine=i686-pc
+	  os=-nacl
+	  ;;
 	# Object if more than one company name word.
 	*-*-*)
 		echo Invalid configuration \`$1\': machine \`$basic_machine\' not recognized 1>&2
@@ -1231,6 +1239,9 @@
 			;;
 		esac
 		;;
+	-nacl*)
+		os=-nacl
+		;;
 	-nto-qnx*)
 		;;
 	-nto*)
