diff --git a/src/TextAction.c b/src/TextAction.c
--- a/src/TextAction.c
+++ b/src/TextAction.c
@@ -47,6 +47,10 @@ in this Software without prior written authorization from The Open Group.
 #include "Private.h"
 #include "XawI18n.h"
 
+#if defined(__native_client__)
+# include <sys/select.h>
+#endif
+
 #define SrcScan			XawTextSourceScan
 #define FindDist		XawTextSinkFindDistance
 #define FindPos			XawTextSinkFindPosition
