diff --git a/Makefile.in b/Makefile.in
index c21ec67..1a5d61d 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -184,7 +184,7 @@ CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
 CYGPATH_W = @CYGPATH_W@
 DEFAULT_POSIX2_VERSION = @DEFAULT_POSIX2_VERSION@
-DEFS = @DEFS@
+DEFS = @DEFS@ -Dmain=nacl_main
 DEPDIR = @DEPDIR@
 DF_PROG = @DF_PROG@
 ECHO_C = @ECHO_C@
@@ -210,7 +210,7 @@ LDFLAGS = @LDFLAGS@
 LIBICONV = @LIBICONV@
 LIBINTL = @LIBINTL@
 LIBOBJS = @LIBOBJS@
-LIBS = @LIBS@
+LIBS = @LIBS@ $(EXTRA_LIBS)
 LIB_CLOCK_GETTIME = @LIB_CLOCK_GETTIME@
 LIB_CRYPT = @LIB_CRYPT@
 LIB_NANOSLEEP = @LIB_NANOSLEEP@
diff --git a/lib/mktime.c b/lib/mktime.c
index ce3de58..fea6229 100644
--- a/lib/mktime.c
+++ b/lib/mktime.c
@@ -127,8 +127,8 @@ ydhms_diff (long int year1, long int yday1, int hour1, int min1, int sec1,
 	    int year0, int yday0, int hour0, int min0, int sec0)
 {
   verify (C99_integer_division, -1 / 2 == 0);
-  verify (long_int_year_and_yday_are_wide_enough,
-	  INT_MAX <= LONG_MAX / 2 || TIME_T_MAX <= UINT_MAX);
+  /* We removed unnecessary verification. Recent coreutils does not
+   * have this check either. */
 
   /* Compute intervening leap days correctly even if year is negative.
      Take care to avoid integer overflow here.  */
diff --git a/man/Makefile.in b/man/Makefile.in
index 2bfa6f4..30022a9 100644
--- a/man/Makefile.in
+++ b/man/Makefile.in
@@ -143,7 +143,9 @@ DIST_SOURCES =
 man1dir = $(mandir)/man1
 am__installdirs = "$(DESTDIR)$(man1dir)"
 NROFF = nroff
-MANS = $(dist_man_MANS)
+# The build rule for man pages are broken for NaCl, and we do not need
+# man pages anyway.
+#MANS = $(dist_man_MANS)
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 ACLOCAL = @ACLOCAL@
 ALLOCA = @ALLOCA@
diff --git a/src/Makefile.in b/src/Makefile.in
index 2d4a63f..d4f43bf 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -716,7 +716,7 @@ CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
 CYGPATH_W = @CYGPATH_W@
 DEFAULT_POSIX2_VERSION = @DEFAULT_POSIX2_VERSION@
-DEFS = @DEFS@
+DEFS = @DEFS@ -Dmain=nacl_main
 DEPDIR = @DEPDIR@
 DF_PROG = @DF_PROG@
 ECHO_C = @ECHO_C@
@@ -742,7 +742,7 @@ LDFLAGS = @LDFLAGS@
 LIBICONV = @LIBICONV@
 LIBINTL = @LIBINTL@
 LIBOBJS = @LIBOBJS@
-LIBS = @LIBS@
+LIBS = @LIBS@ $(EXTRA_LIBS)
 LIB_CLOCK_GETTIME = @LIB_CLOCK_GETTIME@
 LIB_CRYPT = @LIB_CRYPT@
 LIB_NANOSLEEP = @LIB_NANOSLEEP@
diff --git a/src/fs.h b/src/fs.h
index 675dd61..c2eb64b 100644
--- a/src/fs.h
+++ b/src/fs.h
@@ -2,7 +2,10 @@
    Please send additions to meskes@debian.org.
    This file is generated automatically from ./stat.c. */
 
-#if defined __linux__
+/* nacl_io does not have statfs so the following magic numbers which
+   specify the type of a filesystem is useless, but we need their
+   definitions to compile everything in coreutils. */
+#if defined __linux__ || defined __native_client__
 # define S_MAGIC_AFFS 0xADFF
 # define S_MAGIC_DEVPTS 0x1CD1
 # define S_MAGIC_EXT 0x137D
diff --git a/src/tee.c b/src/tee.c
index 5867539..df6859b 100644
--- a/src/tee.c
+++ b/src/tee.c
@@ -31,6 +31,8 @@
 
 #define AUTHORS "Mike Parker", "Richard M. Stallman", "David MacKenzie"
 
+/* Recent coreutils uses tee_files instead of tee to avoid a conflict. */
+#define tee tee_files
 static int tee (int nfiles, const char **files);
 
 /* If nonzero, append to output files rather than truncating them. */
@@ -158,7 +160,7 @@ main (int argc, char **argv)
    Return 0 if successful, 1 if any errors occur. */
 
 static int
-tee (int nfiles, const char **files)
+tee_files (int nfiles, const char **files)
 {
   FILE **descriptors;
   char buffer[BUFSIZ];
diff --git a/src/test.c b/src/test.c
index ef2b35f..35feef0 100644
--- a/src/test.c
+++ b/src/test.c
@@ -124,7 +124,13 @@ test_syntax_error (char const *format, char const *arg)
 #if HAVE_SETREUID && HAVE_SETREGID
 /* Do the same thing access(2) does, but use the effective uid and gid.  */
 
+/* Remove static as nacl-glibc declares this function as non-static
+ * function but does not provide the implementation. */
+#if defined(__native_client__)
+int
+#else
 static int
+#endif
 eaccess (char const *file, int mode)
 {
   static int have_ids;
