diff --git a/Makefile.SH b/Makefile.SH
--- a/Makefile.SH
+++ b/Makefile.SH
@@ -478,7 +478,11 @@ c3 = taint.c toke.c util.c deb.c run.c universal.c pad.c globals.c keywords.c
 c4 = perlio.c perlapi.c numeric.c mathoms.c locale.c pp_pack.c pp_sort.c caretx.c
 c5 = $(madlysrc) $(mallocsrc)
 
+ifeq  ($(NACL_BUILD), host)
 c = $(c1) $(c2) $(c3) $(c4) $(c5) miniperlmain.c opmini.c perlmini.c
+else
+c = $(c1) $(c2) $(c3) $(c4) $(c5) miniperlmain.c opmini.c perlmini.c perl_pepper.c
+endif
 
 obj0 = op$(OBJ_EXT) perl$(OBJ_EXT)
 obj0mini = perlmini$(OBJ_EXT) opmini$(OBJ_EXT) miniperlmain$(OBJ_EXT)
@@ -489,7 +493,12 @@ obj3 = doop$(OBJ_EXT) doio$(OBJ_EXT) regexec$(OBJ_EXT) utf8$(OBJ_EXT) taint$(OBJ
 minindt_obj = $(obj0mini) $(obj1) $(obj2) $(obj3) $(ARCHOBJS)
 mini_obj = $(minindt_obj) $(MINIDTRACE_O)
 ndt_obj = $(obj0) $(obj1) $(obj2) $(obj3) $(ARCHOBJS)
+
+ifeq  ($(NACL_BUILD), host)
 obj = $(ndt_obj) $(DTRACE_O)
+else
+obj = $(ndt_obj) $(DTRACE_O) perl_pepper$(OBJ_EXT)
+endif
 
 perltoc_pod_prereqs = extra.pods pod/perl5202delta.pod pod/perlapi.pod pod/perlintern.pod pod/perlmodlib.pod pod/perluniprops.pod
 generated_pods = pod/perltoc.pod $(perltoc_pod_prereqs)
diff --git a/Makefile.micro b/Makefile.micro
old mode 100644
new mode 100755
--- a/Makefile.micro
+++ b/Makefile.micro
@@ -1,11 +1,14 @@
 LD = $(CC)
 CCFLAGS = -c
+BUILD_CCFLAGS = -c
 DEFINES = -DPERL_CORE -DPERL_MICRO -DSTANDARD_C -DPERL_USE_SAFE_PUTENV \
 	  -DNO_MATHOMS
 OPTIMIZE = 
 CFLAGS = $(DEFINES) $(OPTIMIZE)
 LDFLAGS = 
+BUILD_LDFLAGS = 
 LIBS = -lm
+BUILD_LIBS = -lm
 _O = .o
 ENV = env
 PERL = perl
@@ -14,7 +17,7 @@ RUN =
 
 all:	microperl
 
-O = uav$(_O) udeb$(_O) udoio$(_O) udoop$(_O) udump$(_O) \
+O = uav$(_O) ucaretx$(_O) udeb$(_O) udoio$(_O) udoop$(_O) udump$(_O) \
 	uglobals$(_O) ugv$(_O) uhv$(_O) umro$(_O)\
 	umg$(_O) uperlmain$(_O) uop$(_O) ureentr$(_O) \
 	upad$(_O) uperl$(_O) uperlio$(_O) uperly$(_O) upp$(_O) \
@@ -66,6 +69,9 @@ perly.h: perly.y
 
 uav$(_O):	$(HE) av.c
 	$(CC) $(CCFLAGS) -o $@ $(CFLAGS) av.c
+	
+ucaretx$(_O):	$(HE) caretx.c
+	$(CC) $(CCFLAGS) -o $@ $(CFLAGS) caretx.c
 
 udeb$(_O):	$(HE) deb.c
 	$(CC) $(CCFLAGS) -o $@ $(CFLAGS) deb.c
@@ -184,10 +190,10 @@ ubitcount.h: ugenerate_uudmap$(_X)
 	$(RUN) ./ugenerate_uudmap$(_X) $(generated_headers)
 
 ugenerate_uudmap$(_O): generate_uudmap.c
-	$(CC) $(CCFLAGS) -o $@ $(CFLAGS) generate_uudmap.c
+	$(BUILD_CC) $(BUILD_CCFLAGS) -o $@ $(CFLAGS) generate_uudmap.c
 
 ugenerate_uudmap$(_X): ugenerate_uudmap$(_O)
-	$(LD) -o ugenerate_uudmap $(LDFLAGS) ugenerate_uudmap$(_O) $(LIBS)
+	$(BUILD_LD) -o ugenerate_uudmap $(BUILD_LDFLAGS) ugenerate_uudmap$(_O) $(BUILD_LIBS)
 
 microtest: microperl
 	- cd t && (rm -f perl; ln -s ../microperl perl) \
diff --git a/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm b/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
old mode 100644
new mode 100755
--- a/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
+++ b/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
@@ -89,11 +89,19 @@ struct perl_vars* Perl_GetVarsPrivate(void) { return my_plvarsp; }
 #ifdef NO_ENV_ARRAY_IN_MAIN
 extern char **environ;
 int
+#ifdef __native_client__
+nacl_perl_main(int argc, char **argv)
+#else
 main(int argc, char **argv)
+#endif
 #else
 int
+#ifdef __native_client__
+nacl_perl_main(int argc, char **argv, char **env)
+#else
 main(int argc, char **argv, char **env)
 #endif
+#endif
 {
     int exitstatus, i;
 #ifdef PERL_GLOBAL_STRUCT
diff --git a/lib/ExtUtils/t/Embed.t b/lib/ExtUtils/t/Embed.t
old mode 100644
new mode 100755
--- a/lib/ExtUtils/t/Embed.t
+++ b/lib/ExtUtils/t/Embed.t
@@ -168,11 +168,20 @@ struct perl_vars* Perl_GetVarsPrivate(void) { return my_plvarsp; }
 #endif
 
 #ifdef NO_ENV_ARRAY_IN_MAIN
+#ifdef __native_client__
+int nacl_perl_main(int argc, char **argv) {
+    char **env;
+#else
 int main(int argc, char **argv) {
     char **env;
+#endif
+#else
+#ifdef __native_client__
+int nacl_perl_main(int argc, char **argv, char **env) {
 #else
 int main(int argc, char **argv, char **env) {
 #endif
+#endif
     PerlInterpreter *my_perl;
 #ifdef PERL_GLOBAL_STRUCT
     struct perl_vars *my_vars = init_global_struct();
diff --git a/x2p/a2py.c b/x2p/a2py.c
old mode 100644
new mode 100755
--- a/x2p/a2py.c
+++ b/x2p/a2py.c
@@ -57,7 +57,11 @@ usage()
 #endif
 
 int
+#ifdef __native_client__
+nacl_main(int argc, const char **argv)
+#else
 main(int argc, const char **argv)
+#endif
 {
     STR *str;
     int i;
