diff -aur libxml2-2.7.8/config.sub libxml2-2.7.8-nacl/config.sub
--- libxml2-2.7.8/config.sub	2010-11-04 10:28:16.000000000 -0700
+++ libxml2-2.7.8-nacl/config.sub	2013-01-23 17:17:31.245197406 -0800
@@ -775,6 +775,10 @@
 		basic_machine=i370-ibm
 		os=-mvs
 		;;
+	nacl*)
+		basic_machine=le32-unknown
+		os=-nacl
+		;;
 	ncr3000)
 		basic_machine=i486-ncr
 		os=-sysv4
@@ -1302,7 +1306,7 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* | -nacl*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -aur libxml2-2.7.8/xmlIO.c libxml2-2.7.8-nacl/xmlIO.c
--- libxml2-2.7.8/xmlIO.c	2010-10-11 23:25:32.000000000 -0700
+++ libxml2-2.7.8-nacl/xmlIO.c	2013-01-23 17:18:59.255645774 -0800
@@ -3649,10 +3649,12 @@
 	else *cur = 0;
 	ret = xmlMemStrdup(dir);
     } else {
+#ifndef __native_client__
         if (getcwd(dir, 1024) != NULL) {
 	    dir[1023] = 0;
 	    ret = xmlMemStrdup(dir);
 	}
+#endif
     }
     return(ret);
 #undef IS_XMLPGD_SEP
