diff -urN pango-1.29.3/config.sub pango-1.29.3-nacl/config.sub
--- pango-1.29.3/config.sub	2011-06-16 12:17:39.000000000 -0700
+++ pango-1.29.3-nacl/config.sub	2011-09-26 15:34:05.719401000 -0700
@@ -319,6 +319,10 @@
 	i*86 | x86_64)
 	  basic_machine=$basic_machine-pc
 	  ;;
+        nacl*)
+          basic_machine=i686-pc
+          os=-nacl
+          ;;
 	# Object if more than one company name word.
 	*-*-*)
 		echo Invalid configuration \`$1\': machine \`$basic_machine\' not recognized 1>&2
@@ -1302,7 +1306,7 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* | -nacl*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
