diff --git a/Makefile.in b/Makefile.in
index 5349180..8e354ee 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -225,20 +225,20 @@ bochs_plugins:
 	@CD_UP_ONE@
 
 bximage@EXE@: misc/bximage.o
-	@LINK_CONSOLE@ $(BXIMAGE_LINK_OPTS) misc/bximage.o
+	@LINK_CONSOLE@ $(NACLBXLIBS) $(BXIMAGE_LINK_OPTS) misc/bximage.o
 
 bxcommit@EXE@: misc/bxcommit.o
-	@LINK_CONSOLE@ misc/bxcommit.o
+	@LINK_CONSOLE@ $(NACLBXLIBS) $(BXIMAGE_LINK_OPTS) misc/bxcommit.o
 
 niclist@EXE@: misc/niclist.o
 	@LINK_CONSOLE@ misc/niclist.o
 
 # compile with console CXXFLAGS, not gui CXXFLAGS
 misc/bximage.o: $(srcdir)/misc/bximage.c $(srcdir)/iodev/hdimage.h
-	$(CC) @DASH@c $(BX_INCDIRS) $(CFLAGS_CONSOLE) $(srcdir)/misc/bximage.c @OFP@$@
+	$(CC) @DASH@c $(BX_INCDIRS) $(CFLAGS_CONSOLE) $(NACLBXLIBS) $(srcdir)/misc/bximage.c @OFP@$@
 
 misc/bxcommit.o: $(srcdir)/misc/bxcommit.c $(srcdir)/iodev/hdimage.h
-	$(CC) @DASH@c $(BX_INCDIRS) $(CFLAGS_CONSOLE) $(srcdir)/misc/bxcommit.c @OFP@$@
+	$(CC) @DASH@c $(BX_INCDIRS) $(CFLAGS_CONSOLE) $(NACLBXLIBS) $(srcdir)/misc/bxcommit.c @OFP@$@
 
 misc/niclist.o: $(srcdir)/misc/niclist.c
 	$(CC) @DASH@c $(BX_INCDIRS) $(CFLAGS_CONSOLE) $(srcdir)/misc/niclist.c @OFP@$@
diff --git a/config.cc b/config.cc
index 5b752d7..fd35bdc 100755
--- a/config.cc
+++ b/config.cc
@@ -292,7 +292,7 @@ void bx_init_options()
       "ips", "Emulated instructions per second (IPS)",
       "Emulated instructions per second, used to calibrate bochs emulated time with wall clock time.",
       BX_MIN_IPS, BX_MAX_BIT32U,
-      4000000);
+      15000000);
 #if BX_SUPPORT_SMP
   new bx_param_num_c(cpu_param,
       "quantum", "Quantum ticks in SMP simulation",
diff --git a/config.sub b/config.sub
index 204218c..5bdb7b7 100755
--- a/config.sub
+++ b/config.sub
@@ -283,6 +283,7 @@ case $basic_machine in
 	| moxie \
 	| mt \
 	| msp430 \
+	| nacl \
 	| nios | nios2 \
 	| ns16k | ns32k \
 	| or32 \
diff --git a/configure b/configure
index 4f28619..d22edad 100755
--- a/configure
+++ b/configure
@@ -3748,11 +3748,11 @@ $as_echo "no" >&6; }
 $as_echo "$as_me: failed program was:" >&5
 sed 's/^/| /' conftest.$ac_ext >&5
 
-{ { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
-$as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
-{ as_fn_set_status 77
-as_fn_error "C compiler cannot create executables
-See \`config.log' for more details." "$LINENO" 5; }; }
+#{ { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
+#$as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
+#{ as_fn_set_status 77
+#as_fn_error "C compiler cannot create executables
+#See \`config.log' for more details." "$LINENO" 5; }; }
 else
   { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
 $as_echo "yes" >&6; }
@@ -3792,10 +3792,11 @@ for ac_file in conftest.exe conftest conftest.*; do
   esac
 done
 else
-  { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
-$as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
-as_fn_error "cannot compute suffix of executables: cannot compile and link
-See \`config.log' for more details." "$LINENO" 5; }
+   echo "skipping suffix of executables check"
+#  { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
+#$as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
+#as_fn_error "cannot compute suffix of executables: cannot compile and link
+#See \`config.log' for more details." "$LINENO" 5; }
 fi
 rm -f conftest conftest$ac_cv_exeext
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_exeext" >&5
@@ -24340,6 +24341,10 @@ fi
 DIALOG_OBJS=""
 EXPORT_DYNAMIC="-export-dynamic"
 case $target in
+  *nacl*)
+    # nacl does not support the plugin mechanism
+    EXPORT_DYNAMIC=
+  ;;
   *-pc-windows* | *-pc-winnt* | *-cygwin* | *-mingw*)
     if test "$bx_plugins" = 1; then
       { $as_echo "$as_me:${as_lineno-$LINENO}: checking for opendir in -lmingwex" >&5
diff --git a/cpu/Makefile.in b/cpu/Makefile.in
index a6541fd..4d657e9 100644
--- a/cpu/Makefile.in
+++ b/cpu/Makefile.in
@@ -35,6 +35,8 @@ X_LIBS = @X_LIBS@
 X_PRE_LIBS = @X_PRE_LIBS@
 RANLIB = @RANLIB@
 
+LIBS += $(NACLBXLIBS)
+
 
 
 BX_INCDIRS = -I.. -I$(srcdir)/.. -I../@INSTRUMENT_DIR@ -I$(srcdir)/../@INSTRUMENT_DIR@
diff --git a/cpu/init.cc b/cpu/init.cc
index 3568f79..0883c8d 100644
--- a/cpu/init.cc
+++ b/cpu/init.cc
@@ -160,6 +160,7 @@ void BX_CPU_C::initialize(void)
     BX_CPU_THIS_PTR msrs[n] = 0;
   }
   const char *msrs_filename = SIM->get_param_string(BXPN_CONFIGURABLE_MSRS_PATH)->getptr();
+  msrs_filename = "/msrs/msrs.def";
   load_MSRs(msrs_filename);
 #endif
 
diff --git a/iodev/Makefile.in b/iodev/Makefile.in
index 8ef3da7..0117a96 100644
--- a/iodev/Makefile.in
+++ b/iodev/Makefile.in
@@ -56,7 +56,7 @@ USBDEV_OBJS = @USBDEV_OBJS@
 SCSI_OBJS = @SCSI_OBJS@
 
 BX_INCDIRS = -I.. -I$(srcdir)/.. -I../@INSTRUMENT_DIR@ -I$(srcdir)/../@INSTRUMENT_DIR@
-LOCAL_CXXFLAGS = $(MCH_CFLAGS)
+LOCAL_CXXFLAGS = $(MCH_CFLAGS) -D__nativeclient__=1
 
 PCIDEV_CXXFLAGS = -I$(srcdir)/../host/linux/pcidev
 
diff --git a/iodev/cdrom.cc b/iodev/cdrom.cc
index ab524ac..a46430b 100644
--- a/iodev/cdrom.cc
+++ b/iodev/cdrom.cc
@@ -236,7 +236,7 @@ typedef struct _CDROM_TOC_SESSION_DATA {
 #else // all others (Irix, Tru64)
 #include <sys/types.h>
 #include <sys/stat.h>
-#include <sys/ioctl.h>
+//#include <sys/ioctl.h>
 #define BX_CD_FRAMESIZE 2048
 #define CD_FRAMESIZE 2048
 #endif
diff --git a/iodev/harddrv.cc b/iodev/harddrv.cc
index 18df144..4504862 100644
--- a/iodev/harddrv.cc
+++ b/iodev/harddrv.cc
@@ -307,7 +307,7 @@ void bx_hard_drive_c::init(void)
         BX_HD_THIS channels[channel].drives[device].hdimage->sectors = spt;
 
         /* open hard drive image file */
-        if ((BX_HD_THIS channels[channel].drives[device].hdimage->open(SIM->get_param_string("path", base)->getptr())) < 0) {
+	if ((BX_HD_THIS channels[channel].drives[device].hdimage->open(SIM->get_param_string("path", base)->getptr())) < 0) {
           BX_PANIC(("ata%d-%d: could not open hard drive image file '%s'", channel, device, SIM->get_param_string("path", base)->getptr()));
           return;
         }
@@ -344,7 +344,7 @@ void bx_hard_drive_c::init(void)
             }
           } else {
             if (disk_size != BX_HD_THIS channels[channel].drives[device].hdimage->hd_size) {
-              BX_PANIC(("ata%d-%d disk size doesn't match specified geometry", channel, device));
+              //BX_PANIC(("ata%d-%d disk size doesn't match specified geometry", channel, device));
               // workaround large files problem with diskimages
               BX_HD_THIS channels[channel].drives[device].hdimage->hd_size = disk_size;
             }
diff --git a/iodev/serial.cc b/iodev/serial.cc
index d04beca..226b151 100644
--- a/iodev/serial.cc
+++ b/iodev/serial.cc
@@ -30,6 +30,7 @@
 // is used to know when we are exporting symbols and when we are importing.
 #define BX_PLUGGABLE
 
+#ifndef __nativeclient__
 #ifndef WIN32
 #include <sys/socket.h>
 #include <netinet/in.h>
@@ -37,6 +38,7 @@
 #define closesocket(s)    close(s)
 typedef int SOCKET;
 #endif
+#endif
 
 #include "iodev.h"
 #include "serial.h"
@@ -104,7 +106,6 @@ bx_serial_c::~bx_serial_c(void)
 #endif
           break;
         case BX_SER_MODE_SOCKET:
-          if (BX_SER_THIS s[i].socket_id >= 0) closesocket(BX_SER_THIS s[i].socket_id);
           break;
         case BX_SER_MODE_PIPE:
 #ifdef WIN32
@@ -303,11 +304,8 @@ bx_serial_c::init(void)
         BX_SER_THIS mouse_type = SIM->get_param_enum(BXPN_MOUSE_TYPE)->get();
       } else if (!strncmp(mode, "socket", 6)) {
         BX_SER_THIS s[i].io_mode = BX_SER_MODE_SOCKET;
-        struct sockaddr_in  sin;
-        struct hostent      *hp;
         char                host[BX_PATHNAME_LEN];
         int                 port;
-        SOCKET              socket;
         bx_bool             server = !strcmp(mode, "socket-server");
 
 #if defined(WIN32)
@@ -332,15 +330,19 @@ bx_serial_c::init(void)
         }
         port = atoi(substr);
 
+#ifndef __nativeclient__
         hp = gethostbyname(host);
         if (!hp) {
           BX_PANIC(("com%d: gethostbyname failed (%s)", i+1, host));
         }
 
         memset ((char*) &sin, 0, sizeof (sin));
+#endif
 #if BX_HAVE_SOCKADDR_IN_SIN_LEN
        sin.sin_len = sizeof sin;
 #endif
+
+#ifndef __nativeclient__
         memcpy ((char*) &(sin.sin_addr), hp->h_addr, hp->h_length);
         sin.sin_family = hp->h_addrtype;
         sin.sin_port = htons (port);
@@ -372,12 +374,17 @@ bx_serial_c::init(void)
           socket = (SOCKET) -1;
           BX_INFO(("com%d: connect() failed (host:%s, port:%d)",i+1, host, port));
         }
+#endif
 
+#ifndef __nativeclient__
         BX_SER_THIS s[i].socket_id = socket;
         if (socket > 0)
           BX_INFO(("com%d - inet %s - socket_id: %d, ip:%s, port:%d",
             i+1, server ? "server" : "client", socket, host, port));
-      } else if (!strncmp(mode, "pipe", 4)) {
+#endif
+      }
+
+      else if (!strncmp(mode, "pipe", 4)) {
         if (strlen(dev) > 0) {
 #ifdef WIN32
           bx_bool server = !strcmp(mode, "pipe-server");
@@ -1294,6 +1301,7 @@ bx_serial_c::tx_timer(void)
         BX_INFO(("com%d: write to mouse ignored: 0x%02x", port+1, BX_SER_THIS s[port].tsrbuffer));
         break;
       case BX_SER_MODE_SOCKET:
+#ifndef __nativeclient__
         if (BX_SER_THIS s[port].socket_id >= 0) {
 #ifdef WIN32
           BX_INFO(("attempting to write win32 : %c", BX_SER_THIS s[port].tsrbuffer));
@@ -1304,6 +1312,7 @@ bx_serial_c::tx_timer(void)
                   (bx_ptr_t) & BX_SER_THIS s[port].tsrbuffer, 1);
 #endif
       }
+#endif
       case BX_SER_MODE_PIPE:
 #ifdef WIN32
         if (BX_SER_THIS s[port].pipe) {
diff --git a/iodev/serial.h b/iodev/serial.h
index f3b66dc..fa1dc12 100644
--- a/iodev/serial.h
+++ b/iodev/serial.h
@@ -112,7 +112,6 @@ typedef struct {
 
   int io_mode;
   int tty_id;
-  SOCKET socket_id;
   FILE *output;
 #ifdef WIN32
   HANDLE pipe;
diff --git a/iodev/slowdown_timer.cc b/iodev/slowdown_timer.cc
index 0af0608..f700185 100644
--- a/iodev/slowdown_timer.cc
+++ b/iodev/slowdown_timer.cc
@@ -147,7 +147,7 @@ void bx_slowdown_timer_c::handle_timer()
 #elif BX_HAVE_SLEEP
     sleep(usectosec(s.Q));
 #else
-#error do not know have to sleep
+    sleep(usectosec(s.Q));
 #endif    //delay(wanttime-totaltime);
     /* alternatively: delay(Q);
      * This works okay because we share the delay between
diff --git a/iodev/vvfat.cc b/iodev/vvfat.cc
index 30ceb66..b27b487 100644
--- a/iodev/vvfat.cc
+++ b/iodev/vvfat.cc
@@ -1529,7 +1529,6 @@ bx_bool vvfat_image_t::write_file(const char *path, direntry_t *entry, bx_bool c
   } else {
     ut.actime = ut.modtime;
   }
-  utime(path, &ut);
 #else
   hFile = CreateFile(path, GENERIC_WRITE, 0, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
   if (hFile != INVALID_HANDLE_VALUE) {
diff --git a/logio.cc b/logio.cc
index 09d5c1f..3aff4bb 100644
--- a/logio.cc
+++ b/logio.cc
@@ -25,6 +25,8 @@
 #include "iodev/iodev.h"
 #include <assert.h>
 
+#include "osdep.h"
+
 #if BX_WITH_CARBON
 #define Float32 KLUDGE_Float32
 #define Float64 KLUDGE_Float64
diff --git a/main.cc b/main.cc
index e0c07e1..cad879d 100644
--- a/main.cc
+++ b/main.cc
@@ -475,6 +475,13 @@ int WINAPI WinMain(
 // wxWidgets under win32.
 int CDECL main(int argc, char *argv[])
 {
+#ifdef __native_client__
+  chdir("/img");
+  const char* fake_argv[] = {"bochs", "-qf", "linux-img/bochsrc", NULL};
+  argv = (char**)fake_argv;
+  argc = 3;
+#endif
+
   bx_startup_flags.argc = argc;
   bx_startup_flags.argv = argv;
 #if BX_WITH_SDL && defined(WIN32)
diff --git a/misc/bxcommit.c b/misc/bxcommit.c
index 04d6a84..38c3ea9 100644
--- a/misc/bxcommit.c
+++ b/misc/bxcommit.c
@@ -26,6 +26,8 @@
 
 #include <string.h>
 
+#include "../osdep.h"
+
 #if !BX_HAVE_SNPRINTF
 #include <stdarg.h>
 /* XXX use real snprintf */
@@ -43,11 +45,10 @@ int snprintf(char *s, size_t maxlen, const char *format, ...)
 }
 #endif  /* !BX_HAVE_SNPRINTF */
 
-#include "../osdep.h"
-
 #define HDIMAGE_HEADERS_ONLY 1
 #include "../iodev/hdimage.h"
 
+
 char *EOF_ERR = "ERROR: End of input";
 char *rcsid = "$Id: bxcommit.c,v 1.14 2009/04/14 09:45:22 sshwarts Exp $";
 char *divider = "========================================================================";
diff --git a/osdep.cc b/osdep.cc
index 80aee99..9223241 100644
--- a/osdep.cc
+++ b/osdep.cc
@@ -54,9 +54,9 @@ int bx_snprintf (char *s, size_t maxlen, const char *format, ...)
 #endif  /* !BX_HAVE_SNPRINTF */
 
 #if !BX_HAVE_VSNPRINTF
-int bx_vsnprintf (char *s, size_t maxlen, const char *format, va_list arg)
+int bx_vsnprintf (char *s, size_t maxlen, const char *format, ...)
 {
-  return vsprintf (s, format, arg);
+  return snprintf(s, maxlen, format);
 }
 #endif /* !BX_HAVE_VSNPRINTF*/
 
diff --git a/osdep.h b/osdep.h
index 7699dbd..f72c504 100644
--- a/osdep.h
+++ b/osdep.h
@@ -152,9 +152,16 @@ extern "C" {
   extern int bx_snprintf (char *s, size_t maxlen, const char *format, ...);
 #endif
 
+/* benfits both nnacl and pnacl, avoids tons of warnings from 
+   dubious redefine stricmp -> strcmp  in osdep.h */ 
+#undef BX_HAVE_STRICMP
+#define BX_HAVE_STRICMP 1
+#ifdef __GLIBC__
+#define stricmp strcasecmp
+#endif
 #if !BX_HAVE_VSNPRINTF
   #define vsnprintf bx_vsnprintf
-  extern int bx_vsnprintf (char *s, size_t maxlen, const char *format, va_list arg);
+  extern int bx_vsnprintf (char *s, size_t maxlen, const char *format, ...);
 #endif
 
 #if BX_HAVE_STRTOULL
